src/ga/operators/crossover.rs

```diff
@@ -529,3 +530,126 @@ where
     (child_1, child_2)
   }
 }
+
+/// # PPX crossover operator
+///
+/// This struct implements [CrossoverOperator] trait and can be used with GA.
+///
+/// PPX (Precedence Preservative Crossover), genes are taken in order they appear in parent,
+/// parent is chosen at random, if gene was already taken from other parent then the next un-taken gene
+/// is chosen
+///
+/// P1         : <i>2 4 1 3 5</i> <br>
+/// P2         : <b>5 2 1 4 3</b> <br>
+/// Gene source: 1 1 2 1 2 <br>
+/// Ch         : <i> 2 4 </i> <b> 5 </b> <i> 1<i/> <b> 3</b>
+///
+/// Degenerated case when all genes are taken from the same parent.
+pub struct PPXCrossover<R: Rng> {
+  rng: R,
+}
+
+impl PPXCrossover<ThreadRng> {
+  /// Creates new [PPXCrossover] crossover operator with default RNG
+  pub fn new() -> Self {
+    Self::with_rng(rand::thread_rng())
+  }
+}
+
+impl<R: Rng> PPXCrossover<R> {
+  /// Creates new [PPXCrossover] crossover operator with custom RNG
+  pub fn with_rng(rng: R) -> Self {
+    Self { rng }
+  }
+
+  /// Helper function for [PPXCrossover::apply]
+  /// ## Arguments
+  ///
+  /// * `p1` - First parent to take part in crossover
+  /// * `p2` - Second parent to take part in crossover
+  /// * `take_from_p1` - Which genes should be taken from p1
+  fn create_child<GeneT, ChT>(
+    &self,
+    p1: &Individual<ChT>,
+    p2: &Individual<ChT>,
+    take_from_p1: &[bool],
+  ) -> Individual<ChT>
+  where
+    ChT: Chromosome + Index<usize, Output = GeneT> + Push<GeneT, PushedOut = Nothing>,
+    GeneT: Copy + Eq + Hash,
+  {
+    let chromosome_len = p1.chromosome_ref().len();
+
+    let mut already_taken: HashSet<GeneT> = HashSet::new();
+
+    let mut child: Individual<ChT> = Individual::new();
+    let mut index_p: [usize; 2] = [0, 0];
+    let parents = [p1, p2];
+
+    while child.chromosome_ref().len() < chromosome_len {
+      let index_child = child.chromosome_ref().len();
+      let parent_i: usize = if take_from_p1[index_child] { 0 } else { 1 };
+
+      while child.chromosome_ref().len() == index_child {
+        let gene = parents[parent_i].chromosome_ref()[index_p[parent_i]];
+        index_p[parent_i] += 1;
+
+        if !already_taken.contains(&gene) {
+          already_taken.push(gene);
+          child.chromosome_ref_mut().push(gene);
+        }
+      }
+    }
+
+    child
+  }
+}
+
+impl<GeneT, ChT, R> CrossoverOperator<ChT> for PPXCrossover<R>
```

Comment:
Also please note, that we don't use `Crossover` suffix in any other operator, maybe except `OrderedCrossover`, because I overlooked it in review...

```suggestion
impl<GeneT, ChT, R> CrossoverOperator<ChT> for Ppx<R>
```

Reply:
:ok: